### PR TITLE
feat(profiling): add keyboard navigation to flamegraph

### PR DESCRIPTION
**Pull request by eliashussary**
This PR adds keyboard navigation to the Flamegraph view.

https://user-images.githubusercontent.com/7349258/194380497-b1feb56e-1c1a-4af2-b1e8-7d96e3bbf62b.mov

### CODE REVIEW

File: static/app/components/profiling/flamegraphZoomView.tsx [hunk 1]
```diff
@@ -782,4 +802,112 @@ const Canvas = styled('canvas')`
...
+  ArrowRight: 'right',
+  Tab: 'right',
+  'Shift+Tab': 'left',
+  Enter: 'down',
+  'Shift+Enter': 'up',
+  w: 'up',
+  s: 'down',
+  a: 'left',
+  d: 'right',
+};
```
Comment (eliashussary) [inline static/app/components/profiling/flamegraphZoomView.tsx:890]: stolen from my days as a spreadsheet wizard.

File: static/app/components/profiling/flamegraphZoomView.tsx [hunk 2]
```diff
@@ -782,4 +802,112 @@ const Canvas = styled('canvas')`
   position: absolute;
 `;
 
+function selectNearestFrame(n: FlamegraphFrame, dir: 'up' | 'down' | 'left' | 'right') {
```
Comment (JonasBa) [inline static/app/components/profiling/flamegraphZoomView.tsx:805]: Do you mind adding some comments and explaining briefly what/how we expect this to behave? I can imagine this to have a lot of edge cases which makes it hard to follow.

Comment (eliashussary) [reply static/app/components/profiling/flamegraphZoomView.tsx:805]: Will do; realized i didn't add a test for this either, which i should. it would also help serve as some self documentation

Comment (JonasBa) [reply static/app/components/profiling/flamegraphZoomView.tsx]: Yeah, I would add comments somewhere that explain how we want this to behave (and maybe why). For example, I can think of a few edge cases from the top of my head that I'm not sure how they are handled (least it does not seem obvious)

- current nav is at the top of the stack and there are no children, user presses keydown, do we go to next rootmost frame?
- current nav is at the top of the stack and frame has sibling, user presses keydown, do we go to next sibling?
etc etc, I can think of a few more, so it would be nice to know the exceptions and spec this out.

Also, remember that if users switch to an inverted flamegraph, then top/down keys should be inverted

Comment (eliashussary) [reply static/app/components/profiling/flamegraphZoomView.tsx]: > Also, remember that if users switch to an inverted flamegraph, then top/down keys should be inverted

💀

File: static/app/utils/profiling/flamegraph/selectNearestFrame.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  let node = n;
+
+  if (dir === 'up' && parent) {
+    return parent;
+  }
+
+  const child = n.children?.[0];
+  if (dir === 'down' && child) {
+    return child;
+  }
```
Comment (Zylphrex) [inline static/app/utils/profiling/flamegraph/selectNearestFrame.ts:36]: When the direction is down and there is no child this falls through and behaves like a left. Is this intended?

Comment (eliashussary) [reply static/app/utils/profiling/flamegraph/selectNearestFrame.ts:36]: nope, that's a bug!

Comment (Zylphrex) [reply static/app/utils/profiling/flamegraph/selectNearestFrame.ts]: Can you also add some tests for this too? 😄

Comment (eliashussary) [reply static/app/utils/profiling/flamegraph/selectNearestFrame.ts]: done!

File: static/app/components/profiling/flamegraphZoomView.tsx (hunk 2 above)
Comment (eliashussary) [reply static/app/components/profiling/flamegraphZoomView.tsx]: fixed the inverted flamegraph behavior.

for all other cases where we've reached an end of a boundary, i try to just return the "current" highlighted node. in other words, we don't move and its a no-op.

i feel like that's just going to add a bunch of weird edge cases or bounce the user around too much

Comment (JonasBa) [reply static/app/components/profiling/flamegraphZoomView.tsx]: @eliashussary Since we already allow jumping left-right we should aim on allowing users to use top/down to cycle between stacks too. I don't think implementing cycling should be too hard here (I'd be happy to pair if you want), it would be especially cool because this could allow you to use top/down to jump between stack tops and more easily identify long running stacks.

**Comment by JonasBa** [comment]
On a side note, we never did this, but it would probably make sense for us to have a ghost indicator so that screen readers could properly read the node content

File: static/app/utils/profiling/flamegraph/selectNearestFrame.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+export type DirectionX = 'left' | 'right';
+export type DirectionY = 'up' | 'down';
+export type Direction = DirectionY | DirectionX;
+/**
+ * selectNearestFrame walks a FlamegraphFrame tree the direction specified and returns the nearest
+ * FlamegraphFrame relative to the starting node.
+ *
+ * below is a gist of the walking algorithm:
+ *
+ *  up    => frame's parent
```
[the remark this answers is not in the document]
Comment (Zylphrex) [reply static/app/utils/profiling/flamegraph/selectNearestFrame.ts:12]: why the root of the previous stack? wouldn't that turn up into a left when it's at a root node? I'm not sure if that's intuitive

[the remark this answers is not in the document]
Comment (eliashussary) [reply static/app/utils/profiling/flamegraph/selectNearestFrame.ts:12]: i see what you're saying... i've now realized that this is a bit more borked than i had initially expected.

**Comment by Jesse-Box** [comment]
Have we thought about how the call stack tray responds to certain keyboard navigation? I can imagine it would be useful to press enter on a certain frame and it renders an appropriate call stack.

**Comment by JonasBa** [comment]
> Have we thought about how the call stack tray responds to certain keyboard navigation? I can imagine it would be useful to press enter on a certain frame and it renders an appropriate call stack.

Yeah, that is a good point @Jesse-Box. Right now we do nothing to sync the two, but it may make sense to
 have a shortcut to go from table -> chart and vice versa. I would probably opt to do this on a "pull" basis, similarly to what we have on the table, just in reverse (see screenshot of the ctx menu in the dashboard)
CleanShot 2022-10-19 at 09 28 05@2x

I imagine we could have a similar thing when users open a context menu on the chart itself, wdyt?

File: static/app/utils/profiling/flamegraph/selectNearestFrame.spec.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
+import {selectNearestFrame} from './selectNearestFrame';
+
+class Node {
```
Comment (JonasBa) [inline static/app/utils/profiling/flamegraph/selectNearestFrame.spec.ts:3]: I don't think we need this anymore right - we should be doing this using the flamegraph class or the frame class

File: static/app/utils/profiling/flamegraph/selectNearestFrame.ts [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+export type DirectionX = 'left' | 'right';
+export type DirectionY = 'up' | 'down';
+export type Direction = DirectionY | DirectionX;
+/**
+ * selectNearestFrame walks a FlamegraphFrame tree the direction specified and returns the nearest
+ * FlamegraphFrame relative to the starting node.
+ *
+ * below is a gist of the walking algorithm:
+ *
+ *  up    =>
```
Comment (JonasBa) [inline static/app/utils/profiling/flamegraph/selectNearestFrame.ts:12]: It would probably be easier to follow the logic if this is inlined in the fn body where the different different branching occurs

File: static/app/components/profiling/flamegraphZoomView.tsx [hunk 7]
```diff
@@ -782,4 +808,51 @@ const Canvas = styled('canvas')`
...
+  'Shift+Tab': 'left',
+  Enter: 'down',
+  'Shift+Enter': 'up',
+  w: 'up',
+  s: 'down',
+  a: 'left',
+  d: 'right',
+};
+
+function handleFlamegraphKeyboardNavigation(
```
Comment (JonasBa) [inline static/app/components/profiling/flamegraphZoomView.tsx:827]: Nit, but I wouldn't even dispatch to this fn if the key that is pressed is not a valid direction key, this could allow you to improve the signature a little bit by also checking if currentFrame exists.

File: static/app/components/profiling/flamegraphZoomView.tsx [hunk 8]
```diff
@@ -782,4 +808,51 @@ const Canvas = styled('canvas')`
...
+    return null;
+  }
+
+  const keyboardKey = !inverted
+    ? evt.key
+    : evt.key === 'ArrowUp'
+    ? 'ArrowDown'
+    : evt.key === 'ArrowDown'
+    ? 'ArrowUp'
+    : evt.key;
```
Comment (JonasBa) [inline static/app/components/profiling/flamegraphZoomView.tsx:842]: You can drop negation and it may also be worth adding a comment here to explain as it might be hard for someone without a visual reference of the inversion to understand this

```suggestion
  const keyboardKey = inverted
    ? evt.key === 'ArrowUp' ? 'ArrowDown' : evt.key
    : evt.key === 'ArrowDown' ? 'ArrowUp' : evt.key
    : evt.key;
```

File: static/app/components/profiling/flamegraphZoomView.tsx [hunk 9]
```diff
@@ -782,4 +808,51 @@ const Canvas = styled('canvas')`
...
+
+  const key = evt.shiftKey ? `Shift+${keyboardKey}` : keyboardKey;
+  const direction = keyDirectionMap[key];
+  if (!direction) {
+    return currentFrame;
+  }
+  const nextSelection = selectNearestFrame(currentFrame, direction);
```
Comment (JonasBa) [inline static/app/components/profiling/flamegraphZoomView.tsx:849]: If currentFrame is not falsy, then no matter what key you press, the least you will always get back is the current frame right?

File: static/app/utils/profiling/flamegraph/selectNearestFrame.ts [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    if (!parent) {
+      break;
+    }
+
+    node = parent;
+    parent = parent.parent;
+  }
+  return frame;
```
Comment (JonasBa) [inline static/app/utils/profiling/flamegraph/selectNearestFrame.ts:86]: This is just returning the fn inpu, shouldn't it return `node` instead?

File: static/app/utils/profiling/flamegraph/selectNearestFrame.ts [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    const nextNode = node.children[directionX === 'right' ? 0 : node.children.length - 1];
+    if (!nextNode) {
+      return node;
+    }
+    node = nextNode;
+  }
+
+  return frame;
```
Comment (JonasBa) [inline static/app/utils/profiling/flamegraph/selectNearestFrame.ts:116]: Same return comment here

File: static/app/components/profiling/flamegraphZoomView.tsx (hunk 1 above)
Comment (JonasBa) [reply static/app/components/profiling/flamegraphZoomView.tsx]: I think it's ok, the only one I'm not sure is enter and shift+enter. I'd probably remove those for now and in the future use this to zoom to frame

File: static/app/components/profiling/flamegraphZoomView.tsx (hunk 9 above)
Comment (eliashussary) [reply static/app/components/profiling/flamegraphZoomView.tsx:849]: yeah, if we don't have a keybind set we no-op and return whatever the currentFrame is, if one is already selected.

File: static/app/components/profiling/flamegraphZoomView.tsx (hunk 1 above)
Comment (eliashussary) [reply static/app/components/profiling/flamegraphZoomView.tsx]: i'd drop the enter keybind and make that the default behavior of tab. since we implemented cycling with up/down, we can just make tab cycle by moving down, and shift+tab moving up

File: static/app/components/profiling/flamegraphZoomView.tsx (hunk 8 above)
Comment (eliashussary) [reply static/app/components/profiling/flamegraphZoomView.tsx:842]: your nested ternary code throws a syntax error.

in any event, i found this to be a naive we of handling it. i'm inverting specific keybinds, which is very dumb since any other key bound to up/down directions won't get the proper treatment.

i've remove this in favor of simply inverting `direction` when its either up/down

File: static/app/utils/profiling/flamegraph/selectNearestFrame.ts (hunk 10 above)
Comment (eliashussary) [reply static/app/utils/profiling/flamegraph/selectNearestFrame.ts:86]: we could... i designed it so that it would treat it as a hard boundary. if we were to return node, we'd jump stacks to the target depth.

moving `right` here would bounce me to a frame in the first stack 
image

File: static/app/utils/profiling/flamegraph/selectNearestFrame.ts (hunk 11 above)
Comment (eliashussary) [reply static/app/utils/profiling/flamegraph/selectNearestFrame.ts:116]: technically this return is unreachable, but the linter loses its shit.

File: static/app/components/profiling/flamegraphZoomView.tsx (hunk 8 above)
Comment (JonasBa) [reply static/app/components/profiling/flamegraphZoomView.tsx:842]: >in any event, i found this to be a naive we of handling it. i'm inverting specific keybinds, which is very dumb since any other key bound to up/down directions won't get the proper treatment.

great move (and sorry about the suggestion, wish gh had some linting in the suggestion UI 😅 )

File: static/app/utils/profiling/flamegraph/selectNearestFrame.ts (hunk 11 above)
Comment (JonasBa) [reply static/app/utils/profiling/flamegraph/selectNearestFrame.ts:116]: Ok, then I would propose to break instead of return on L103 and return node. This way the while loop only does the searching. If I'm not mistaken, I think you could rewrite the while loop as something like 
```
while(node && node.depth !== depth){
  node = node.children[directionX === 'right' ? 0 : node.children.length - 1];
}
return node
```

Comment (eliashussary) [reply static/app/utils/profiling/flamegraph/selectNearestFrame.ts:116]: That won't work. `node` can possibly be undefined in this case, breaking its behavior. 

the idea is to find the target depth OR if you've reached the bottom of the branch return that node.

I'd prefer not to change this as i find it easy to follow as is.

**Review by JonasBa — APPROVED** [review]
🚢
